### PR TITLE
Focusscope tree

### PR DESCRIPTION
**Pull request by snowystinger**
Closes https://github.com/adobe/react-spectrum/issues/2444 Closes https://github.com/adobe/react-spectrum/pull/2451

Inspired by https://github.com/adobe/react-spectrum/pull/2445 and the conversation there
Addressed comments here https://github.com/adobe/react-spectrum/pull/3323#issuecomment-1190554875

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

## 📝 Test Instructions:

Make sure focus is restored from popovers/dialogs/actionbar/menus
DialogContainer has a new story added which allows dialogs to open new dialogs indefinitely. Focus should still be restored.

### CODE REVIEW

File: packages/@react-aria/focus/stories/FocusScope.stories.tsx [hunk 1]
```diff
@@ -18,7 +18,7 @@ import ReactDOM from 'react-dom';
 const dialogsRoot = 'dialogsRoot';
 
 interface StoryProps {
-  usePortal: boolean,
+  isPortaled: boolean,
```
Comment (snowystinger) [inline packages/@react-aria/focus/stories/FocusScope.stories.tsx:21]: moved away from usePortal because that sounds like a hook and this is just a boolean prop

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 2]
```diff
@@ -611,3 +620,106 @@ function last(walker: TreeWalker) {
...
+class Tree {
+  private root: Node;
+  private fastMap = new Map<ScopeRef, Node>();
+
+  constructor() {
+    this.root = new Node({scopeRef: null});
+    this.fastMap.set(null, this.root);
+  }
+
+  get size() {
```
Comment (snowystinger) [inline packages/@react-aria/focus/src/FocusScope.tsx:634]: i only use this for testing purposes to make to make sure the tree is cleaned up, thoughts?

**Review by majornista — CHANGES_REQUESTED** [review]

**Comment by majornista** [comment]
This does not yet solve for the use cases and tests added in #2445.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 3]
```diff
@@ -611,3 +620,106 @@ function last(walker: TreeWalker) {
...
+  removeNode(scopeRef: ScopeRef) {
+    // never remove the root
+    if (scopeRef === null) {
+      return;
+    }
+    let node = this.fastMap.get(scopeRef);
+    let parentNode = node.parent;
+    let children = node.children;
+    parentNode.removeChild(node);
+    if (children.length > 0) {
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:662]: ah, thanks for the reminder about that case, I'm surprised a test didn't fail

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 4]
```diff
@@ -104,26 +100,34 @@ export function FocusScope(props: FocusScopeProps) {
...
-        activeScope = parentScope;
-      }
-      scopes.delete(scopeRef);
-    };
-  }, [scopeRef, parentScope]);
+  // add to the focus scope tree in render order because useEffects/useLayoutEffects run children first whereas render runs parent first
+  // which matters when constructing a tree
+  if (focusScopeTree.getNode(parentScope) && !focusScopeTree.getNode(scopeRef)) {
+    focusScopeTree.addNode(scopeRef, parentScope);
+  }
```
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:107]: yeah, i was debating myself over the naming last night, I think you're right, it's confusing as it currently is

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (majornista) [reply packages/@react-aria/focus/src/FocusScope.tsx:662]: I'm not sure the test is there. It's in DialogContainer.

**Comment by majornista** [comment]
Tests and storybook examples from #2445 that need to pass or work in addition to the added FocusScope tests:

1. [DialogContainer: Nested Dialog Containers](https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/dialogcontainer--nested-dialog-containers&providerSwitcher-toastPosition=bottom) from [Storybook Example](https://github.com/adobe/react-spectrum/pull/2445/files#diff-a107f8890804c2c4c6068f3a3b9f9028b2055a7001f398c7046237c1684645abR65-R96) and [test](https://github.com/adobe/react-spectrum/pull/2445/files#diff-039c81f59042d1f629ade1499b47727938f36264cb88ca6f02fcea848eea37bfR187-R261).
2. [Modal: unmounting trigger](https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/modal--unmounting-trigger&providerSwitcher-toastPosition=bottom)
3. [DialogContainer: in a menu](https://reactspectrum.blob.core.windows.net/reactspectrum/4b09194340f607fab1c56739c71f99e8a7d58337/storybook/index.html?path=/story/dialogcontainer--in-a-menu&providerSwitcher-toastPosition=bottom)

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 5]
```diff
@@ -611,3 +620,106 @@ function last(walker: TreeWalker) {
...
+    }
+  }
+
+  removeNode(scopeRef: ScopeRef) {
+    // never remove the root
+    if (scopeRef === null) {
+      return;
+    }
+    let node = this.fastMap.get(scopeRef);
+    let parentNode = node.parent;
```
Comment (majornista) [inline packages/@react-aria/focus/src/FocusScope.tsx:659]: Try this:

```suggestion
    let parentNode = node.parent;
    parentNode.children.forEach(
      child => {
        if (
          child !== node &&
          node.nodeToRestore &&
          child.nodeToRestore &&
          node.scopeRef.current &&
          isElementInScope(child.nodeToRestore, node.scopeRef.current)
        ) {
          child.nodeToRestore = node.nodeToRestore;
        }
      }
    );
```

**Review by majornista — CHANGES_REQUESTED** [review]

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 5 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:659]: This does not work for me, did it work for you? I would be surprised if it did with the CRUD table, what is it solving for?
it might be easier to think through if it was written like
```
    let parentNode = node.parent;
    parentNode.children.forEach(
      sibling => {
        if (
          sibling !== node &&
          node.nodeToRestore &&
          sibling.nodeToRestore &&
          node.scopeRef.current &&
          isElementInScope(sibling.nodeToRestore, node.scopeRef.current)
        ) {
          sibling.nodeToRestore = node.nodeToRestore;
        }
      }
    );
```

Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:659]: hmmm actually, that works in a live example, but not in unit tests

**Review by majornista — CHANGES_REQUESTED** [review]
Some small suggestions.  

But, I'm also curious as to whether the FocusTree can help restoring focus with an ActionBar use case that I'm working to resolve with #2451. An ActionBar has a FocusScope with `restoreFocus` and `contain={false}` that contains an `ActionGroup` and the `ActionButton` to close the `ActionBar`. In the [ActionBar storybook example](https://reactspectrum.blob.core.windows.net/reactspectrum/33c79adc9bb94bd5a6a7a3173cc61680bed9fa20/storybook/index.html?path=/story/actionbar--default&providerSwitcher-toastPosition=bottom), pressing the Escape key to clear the selection and with focus on the close `ActionButton` will close the `ActionBar` and restore focus to the element that had focus before the `ActionBar` opened, however with focus on one of the `ActionGroup` items within the `ActionBar`, pressing Escape fails to restore focus to the appropriate element.

File: packages/@react-spectrum/picker/test/Picker.test.js [hunk 6]
```diff
@@ -576,14 +575,15 @@ describe('Picker', function () {
...
+      act(() => jest.runAllTimers());
 
       expect(listbox).not.toBeInTheDocument();
       expect(picker).toHaveAttribute('aria-expanded', 'false');
       expect(picker).not.toHaveAttribute('aria-controls');
       expect(onOpenChange).toBeCalledTimes(2);
       expect(onOpenChange).toHaveBeenCalledWith(false);
 
-      expect(document.activeElement).not.toBe(picker);
+      expect(document.activeElement).toBe(picker);
```
Comment (snowystinger) [inline packages/@react-spectrum/picker/test/Picker.test.js:586]: I believe this was in error, focus should have been restored because nothing else was on the page

**Review by majornista — APPROVED** [review]

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 7]
```diff
@@ -402,9 +399,34 @@ function useRestoreFocus(scopeRef: RefObject<Element[]>, restoreFocus: boolean,
...
+    scope.forEach(element => element.addEventListener('focusin', onFocus, false));
+    return () => {
+      document.removeEventListener('focusin', onFocus, false);
+      scope.forEach(element => element.removeEventListener('focusin', onFocus, false));
+    };
+  }, [scopeRef, contain]);
+
   // useLayoutEffect instead of useEffect so the active element is saved synchronously instead of asynchronously.
   useLayoutEffect(() => {
     let nodeToRestore = nodeToRestoreRef.current;
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:428]: Why do we still need a local version, and a ref version rather than using the one stored in the tree node?

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 8]
```diff
@@ -470,11 +493,29 @@ function useRestoreFocus(scopeRef: RefObject<Element[]>, restoreFocus: boolean,
...
       }
 
-      if (restoreFocus && nodeToRestore && isElementInScope(document.activeElement, scopeRef.current)) {
+      // if we already lost focus to the body and this was the active scope, then we should attempt to restore
+      if (
+        restoreFocus
+        && nodeToRestore
+        && (
+          isElementInScope(document.activeElement, scopeRef.current)
+          || (document.activeElement === document.body && activeScope === scopeRef)
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:502]: If focus was lost to the body, wouldn't (shouldn't?) activeScope already be reset to `null`?

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 9]
```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
   } while (last);
   return next;
 }
+
+
+class Tree {
+  private root: TreeNode;
+  private fastMap = new Map<ScopeRef, TreeNode>();
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:659]: You could probably get rid of this map if you returned nodes from `addTreeNode` and passed those around (e.g. through context, to hooks, etc.) rather than only the ref object. Then you wouldn't need to look up nodes from scopes, you'd have the nodes which contain the scope instead.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 10]
```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
...
+  removeTreeNode(scopeRef: ScopeRef) {
+    // never remove the root
+    if (scopeRef === null) {
+      return;
+    }
+    let node = this.fastMap.get(scopeRef);
+    let parentNode = node.parent;
+    // when we remove a scope, check if any sibling scopes are trying to restore focus to something inside the scope we're removing
+    // if we are, then replace the siblings restore with the restore from the scope we're removing
+    parentNode.children.forEach(
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:694]: Couldn't `parentNode` be null here?

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 11]
```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
...
+    );
+    let children = node.children;
+    parentNode.removeChild(node);
+    if (children.length > 0) {
+      children.forEach(child => parentNode.addChild(child));
+    }
+    this.fastMap.delete(node.scopeRef);
+  }
+
+  * traversePreOrderDF(node: TreeNode = this.root): Generator<TreeNode> {
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:715]:
```suggestion
  *traversePreOrderDF(node: TreeNode = this.root): Generator<TreeNode> {
```

😂 

Also I think you could simplify the name to just be `traverse` since there are no other traversal methods, and add a comment indicating that it's a pre-order depth first traversal. Otherwise you might need to expand out what `DF` is for those who don't know. I think that's better done in a comment than a really long method name.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 12]
```diff
@@ -470,11 +493,29 @@ function useRestoreFocus(scopeRef: RefObject<Element[]>, restoreFocus: boolean,
...
+      if (
+        restoreFocus
+        && nodeToRestore
+        && (
+          isElementInScope(document.activeElement, scopeRef.current)
+          || (document.activeElement === document.body && activeScope === scopeRef)
+        )
+      ) {
+        // freeze the focusScopeTree so it persists after the raf, otherwise during unmount nodes are removed from it
+        let clonedTree = focusScopeTree.clone();
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:506]: An alternative to cloning the whole tree would be to only clone the path from `nodeToRestore` to the root here, since that's all you care about.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 13]
```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
...
+    let newTree = new Tree();
+    for (let node of this.traversePreOrderDF()) {
+      newTree.addTreeNode(node.scopeRef, node.parent.scopeRef, node.nodeToRestore);
+    }
+    return newTree;
+  }
+}
+
+class TreeNode {
+  private _scopeRef: ScopeRef;
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:736]: underscores aren't necessary because `private`? also this isn't publicly exposed so probably not a big deal.

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 14]
```diff
@@ -611,3 +652,121 @@ function last(walker: TreeWalker) {
...
+  }
+  set nodeToRestore(node: FocusableElement) {
+    this._nodeToRestore = node;
+  }
+  set parent(node: TreeNode) {
+    this._parent = node;
+  }
+  get parent(): TreeNode {
+    return this._parent;
+  }
```
Comment (devongovett) [inline packages/@react-aria/focus/src/FocusScope.tsx:758]: Why are all these getters and setters needed rather than just making the properties public?

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js [hunk 15]
```diff
@@ -175,7 +166,45 @@ describe('DialogContainer', function () {
 
     triggerPress(button);
     act(() => {jest.runAllTimers();});
+    act(() => {jest.runAllTimers();});
 
     expect(queryByRole('dialog')).toBeNull();
   });
+
+  it('should be able to have dialogs open dialogs and still restore focus', function () {
```
Comment (devongovett) [inline packages/@react-spectrum/dialog/test/DialogContainer.test.js:174]:
```suggestion
  it('should be able to have menus open dialogs and still restore focus', function () {
```

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 7 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:428]: just missed replacing it, good catch

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 9 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:659]: hmmm that's a larger change, that'd mean changing activeScope to a TreeNode and passing a TreeNode through context in place of parent
I'm fine doing that but it might be harder to review? Maybe followup?

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 8 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:503]: We don't set active scope when we lose focus to the body, we could start tracking that though, but it might be complex because it'd really just be saying that the root is the active scope for anything not in an actual FocusScope, which is different than saying it's on document.body

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 12 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:507]: true, but i can't imagine there will be that many focus scopes on the screen that this would ever be a performance issue.
still, happy to change it

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 10 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:695]: No, we early return if we're at the root, so the parentNode will always exist

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 11 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx]: haha, sure, that's fine

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 13 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx]: yeah, that was the only reason, i usually just set up most things as private at first, I'm fine having them be public internally though

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 14 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:745]: same as above

File: packages/@react-spectrum/dialog/test/DialogContainer.test.js (hunk 15 above)
Comment (snowystinger) [reply packages/@react-spectrum/dialog/test/DialogContainer.test.js:174]: this is correct, the menu opens a dialog but the dialog opens another dialog

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 16]
```diff
@@ -402,9 +399,34 @@ function useRestoreFocus(scopeRef: RefObject<Element[]>, restoreFocus: boolean,
   // create a ref during render instead of useLayoutEffect so the active element is saved before a child with autoFocus=true mounts.
   const nodeToRestoreRef = useRef(typeof document !== 'undefined' ? document.activeElement as FocusableElement : null);
 
+  // restoring scopes should all track if they are active regardless of contain, but contain already tracks it plus logic to contain the focus
+  // restoring-non-containing scopes should only care if they become active so they can perform the restore
+  useLayoutEffect(() => {
+    let scope = scopeRef.current;
+    if (!restoreFocus || contain) {
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:419]: hmm, yeah. it would need to check if it is the inner-most scope with restoreFocus.

reason I was thinking was if a focus scope started out with both contain and restoreFocus disabled, and then got one of them enabled later, it wouldn't become the active scope if it already contained focus.

but if it's a big change, then we could do it separately (or when it comes up)

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 9 above)
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:673]: sure, might be good followup. I think it could potentially clean up the code a little

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 10 above)
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:708]: you return if you are the root, not if the _parent_ is the root right?

File: packages/@react-aria/focus/src/FocusScope.tsx [hunk 17]
```diff
@@ -402,9 +399,34 @@ function useRestoreFocus(scopeRef: RefObject<Element[]>, restoreFocus: boolean,
...
+      }
+    };
+
+    document.addEventListener('focusin', onFocus, false);
+    scope.forEach(element => element.addEventListener('focusin', onFocus, false));
+    return () => {
+      document.removeEventListener('focusin', onFocus, false);
+      scope.forEach(element => element.removeEventListener('focusin', onFocus, false));
+    };
+  }, [scopeRef, contain]);
```
[the remark this answers is not in the document]
Comment (devongovett) [reply packages/@react-aria/focus/src/FocusScope.tsx:437]: what do you think about this one?

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:419]: yeah, i mean, this is also an issue when you have
```
<FocusScope restoreFocus>
  <FocusScope contain />
</FocusScope>
```

so maybe I should handle it here, but am happy to do follow up
maybe if an active focusscope unmounts, if it's not a restoring focusscope, it marks its parent as the active scope?

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:437]: sorry, was same concern as https://github.com/adobe/react-spectrum/pull/3323#discussion_r931635961

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 8 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:516]: that's a great point, this is going to be a weird race potentially, we should focus on this when we talk this afternoon

File: packages/@react-aria/focus/src/FocusScope.tsx (hunk 10 above)
Comment (snowystinger) [reply packages/@react-aria/focus/src/FocusScope.tsx:708]: right, but the root isn't null, it just has null for the data. root is still a TreeNode

**Review by devongovett — APPROVED** [review]

**Review by devongovett — APPROVED** [review]

**Review by LFDanLu — APPROVED** [review]

**Comment by frontsideair** [comment]
Hi @snowystinger, I'm using Dialog and Menu from the react-aria docs nearly verbatim, but still having this issue. Is there something that I should be doing additionally that's not yet documented for focus to go back to the menu trigger on dialog close?

I can provide code samples or better yet a sandbox if necessary, I just wanted to know if there's an additional step to be done or if it should be fixed out-of-the-box. React Spectrum storybook shows that the issue is fixed but I haven't found a similar working example for react-aria.

**Comment by snowystinger** [comment]
Hey @frontsideair , sorry to hear you're having trouble following the docs. Let's see if we can get your code working.
Any chance you have your work in a codesandbox where I could look at it? If not, here's a starting template https://codesandbox.io/s/wizardly-clarke-nb86fj

Something else to compare to might be the styled example we have in another codesandbox https://react-spectrum.adobe.com/react-aria/useMenu.html#styled-examples I just tested it and focus is restored to the menu trigger in that example. You can also always look at our implementation of the components as well, though I know this is a more complex one, and our components are broken up for reusability. So I'll understand if you don't want to go combing through ours https://github.com/adobe/react-spectrum/tree/main/packages/%40react-spectrum/menu/src

The one thing I know caused me some troubles at some point was a closing animation. Though I'm not quite recalling if that was in relation to this or something else. When I figure out which it was, I'll try to remember to update here. Otherwise, let's start with your code and compare it to what we've got, and see what differences we have.

**Comment by frontsideair** [comment]
Thanks for the quick reply! I tried to check the React Spectrum implementation for reference and as you said it was a bit hard to follow, considering it's relatively easier to create a sandbox from scratch.

I also checked the styled example, and while it works for the simple case, it didn't restore focus when a dialog that was triggered from a menu item was closed. 

Here's a quick repro sandbox that I created from Dialog and Menu styled examples, and I can confirm that focus is not restored to the trigger: https://codesandbox.io/s/sad-platform-58uy0k?file=/src/App.tsx

**Comment by snowystinger** [comment]
O, I see. Yes, this case is quite annoying. Don't worry, it's nothing you've done. We addressed this case in https://github.com/adobe/react-spectrum/pull/3385

It hasn't made it into a release yet. We're gearing up for our next release soon, it will be included then. You're welcome to early test it to make sure we solved the issue. That'd be a great help.

**Comment by frontsideair** [comment]
I updated the sandbox to the nightlies of react-aria and react-stately, but it doesn't seem like fixed. I can't tell if the PR is included in the nightly though, then I have to do a local repro. The last few nightlies seem to be failing though.

**Comment by snowystinger** [comment]
Hmm... I'll check tomorrow. Thanks for trying it out.

**Comment by frontsideair** [comment]
I tried the new release and the issue seems to persist, I updated the sandbox. Also I removed `CSSTransition` to rule it out. I wonder if I constructed the test case wrong, I copied the files from the two official styled examples but some imports were changed so there may potentially be some API changes that didn't make it to the sandbox. The reason I doubt it is because the release announcement example on Twitter seems to be working nicely.

**Comment by snowystinger** [comment]
Thanks for following up. I'll open an issue for it so we don't lose track of this.

**Comment by majornista** [comment]
> Hi @snowystinger, I'm using Dialog and Menu from the react-aria docs nearly verbatim, but still having this issue. Is there something that I should be doing additionally that's not yet documented for focus to go back to the menu trigger on dialog close?
> 
> I can provide code samples or better yet a sandbox if necessary, I just wanted to know if there's an additional step to be done or if it should be fixed out-of-the-box. React Spectrum storybook shows that the issue is fixed but I haven't found a similar working example for react-aria.

See https://github.com/adobe/react-spectrum/issues/3601#issuecomment-1278107917
